src/main/java/com/adobe/target/edge/client/service/DefaultTargetService.java

```diff
@@ -56,23 +62,55 @@ public DefaultTargetService(ClientConfig clientConfig) {
     this.targetHttpClient.addDefaultHeader(SDK_USER_KEY, SDK_USER_VALUE);
     this.targetHttpClient.addDefaultHeader(SDK_VERSION_KEY, SDK_VERSION);
     this.clientConfig = clientConfig;
+    this.telemetryService = new TelemetryService(clientConfig);
   }
 
   @Override
   public TargetDeliveryResponse executeRequest(TargetDeliveryRequest deliveryRequest) {
+    TimingTool timer = new TimingTool();
+    timer.timeStart(TIMING_EXECUTE_REQUEST);
+
     String url = clientConfig.getUrl(getBestLocationHint(deliveryRequest));
+    TargetDeliveryResponse targetDeliveryResponse;
+    synchronized (this) {
+      if(!storedTelemetries.isEmpty()) {
+        Telemetry telemetry = new Telemetry();
+        deliveryRequest.getDeliveryRequest().setTelemetry(telemetry);
+        while (!storedTelemetries.isEmpty()) {
+          telemetry.addEntriesItem(storedTelemetries.poll());
+        }
+      }
+    }
+
     HttpResponse<DeliveryResponse> response =
         targetHttpClient.execute(
             getQueryParams(deliveryRequest),
             url,
             deliveryRequest.getDeliveryRequest(),
             DeliveryResponse.class);
-    return getTargetDeliveryResponse(deliveryRequest, response);
+    targetDeliveryResponse = getTargetDeliveryResponse(deliveryRequest, response);
+
+    /* capture Telemetry information once original request's response is received */
+    addTelemetry(deliveryRequest, timer, targetDeliveryResponse);
+
+    return targetDeliveryResponse;
   }
 
   @Override
   public CompletableFuture<TargetDeliveryResponse> executeRequestAsync(
       TargetDeliveryRequest deliveryRequest) {
+    TimingTool timer = new TimingTool();
+    timer.timeStart(TIMING_EXECUTE_REQUEST);
+
+    synchronized (this) {
```

Comment:
Same as above, we should just be calling `telemetryService.getTelemetry();` here and move the telemetry-specific internals to `TelemetryService`